src/util/fs-normalized.js

```diff
@@ -0,0 +1,170 @@
+
+/**
+ * Unlinks the destination to force a recreation. This is needed on case-insensitive file systems
+ * to force the correct naming when the filename has changed only in character-casing. (Jest -> jest).
+ */
+export const copyFile = async function(data: CopyFileAction, cleanup: () => mixed): Promise<void> {
+  try {
+    await unlink(data.dest);
+    await copyFilePoly(data.src, data.dest, 0, data);
+  } finally {
+    if (cleanup) {
+      cleanup();
+    }
+  }
+};
+
+// Node 8.5.0 introduced `fs.copyFile` which is much faster, so use that when available.
+// Otherwise we fall back to reading and writing files as buffers.
+const copyFilePoly: (src: string, dest: string, flags: number, data: CopyFileAction) => Promise<void> = (
+  src,
+  dest,
+  flags,
+  data,
+) => {
+  if (fs.copyFile) {
+    return new Promise((resolve, reject) =>
+      fs.copyFile(src, dest, flags, err => {
+        if (err) {
+          reject(err);
+        } else {
+          fixTimes(undefined, dest, data).then(() => resolve()).catch(ex => reject(ex));
+        }
+      }),
+    );
+  } else {
+    return copyWithBuffer(src, dest, flags, data);
+  }
+};
+
+const copyWithBuffer: (src: string, dest: string, flags: number, data: CopyFileAction) => Promise<void> = async (
+  src,
+  dest,
+  flags,
+  data,
+) => {
+  // Use open -> write -> futimes -> close sequence to avoid opening the file twice:
+  // one with writeFile and one with utimes
+  const fd = await open(dest, 'w', data.mode);
+  try {
+    const buffer = await readFileBuffer(src);
+    await write(fd, buffer, 0, buffer.length);
+    await fixTimes(fd, dest, data);
+  } finally {
+    await close(fd);
+  }
+};
+
+// We want to preserve file timestamps when copying a file, since yarn uses them to decide if a file has
+// changed compared to the cache.
+// There are some OS specific cases here:
+// * On linux, fs.copyFile does not preserve timestamps, but does on OSX and Win.
+// * On windows, you must open a file with write permissions to call `fs.futimes`.
+// * On OSX you can open with read permissions and still call `fs.futimes`.
+async function fixTimes(fd: ?number, dest: string, data: CopyFileAction): Promise<void> {
+  const doOpen = fd === undefined;
+  let openfd: number = fd ? fd : -1;
+
+  if (disableTimestampCorrection === undefined) {
+    // if timestamps match already, no correction is needed.
+    // the need to correct timestamps varies based on OS and node versions.
+    const destStat = await lstat(dest);
+    disableTimestampCorrection = fileDatesEqual(destStat.mtime, data.mtime);
+  }
+
+  if (disableTimestampCorrection) {
+    return;
+  }
+
+  if (doOpen) {
+    try {
+      openfd = await open(dest, 'a', data.mode);
+    } catch (er) {
+      // file is likely read-only
+      try {
+        openfd = await open(dest, 'r', data.mode);
+      } catch (err) {
+        // We can't even open this file for reading.
+        return;
```

Comment:
This is catching `er` and `err` without logging or throwing anything here.
Issues with this may be difficult to debug.

Reply:
There shouldn't be any issues if it fails. We just won't correct the timestamps (because we can't) and yarn will behave the same way it does now. The comment at the top of the function tries to explain this, but maybe I didn't do a great job of explaining it.

Reply:
It is not critical, but if this code is not working on some machine it will look like a fairly big slowdown without a good way to debug it.

Reply:
@rally25rs, I agree with @james-rabbit, it makes sense to put verbose logs that `open -x` failed for the future when we might be debugging why install on one OS is slower than the other.
Would you add it?